src/handlers/mod.rs

```diff
@@ -338,41 +342,87 @@ pub async fn list_blobs_handler(
     Json(blob_descriptors).into_response()
 }
 
-pub async fn upload_blob_handler(
-    State(app_state): State<AppState>,
-    AuthHeader(auth_event): AuthHeader,
-    headers: HeaderMap,
-    body: Bytes,
-) -> impl IntoResponse {
+/// Perform checks for upload that can be done without the actual blob, like whether user is
+/// allowed to upload files.
+/// Returns the validated autentication event and content type on success, or a HTTP status code
+/// and message on failure.
+pub fn upload_blob_prechecks(
+    app_state: &AppState,
+    auth_event: Option<Event>,
+    headers: &HeaderMap,
+) -> Result<(Event, Option<String>), (StatusCode, String)> {
     // Validate that uploads are enabled
     if !app_state.config.upload.enabled {
-        let json = Json(ErrorResponse {
-            message: "Uploads are disabled".to_string(),
-        });
-        return (StatusCode::NOT_FOUND, json).into_response();
+        return Err((StatusCode::NOT_FOUND, "Uploads are disabled".to_string()));
     }
 
     // Get the auth event from HTTP headers
     let auth_event = match auth_event {
         Some(auth_event) => auth_event,
         None => {
-            let json = Json(ErrorResponse {
-                message: "Authorization event required to upload a blob".to_string(),
-            });
-            return (StatusCode::UNAUTHORIZED, json).into_response();
+            return Err((StatusCode::UNAUTHORIZED, "Authorization event required to upload a blob".to_string()));
         }
     };
 
     // Validate the authorization event
     match validate_auth_event(&auth_event, "upload") {
         Ok(_) => {}
         Err(error_msg) => {
+            return Err((StatusCode::UNAUTHORIZED, error_msg.to_string()));
+        }
+    }
+
+    // Check if public key is allowed to upload to the server
+    if let Err(e) = is_public_key_allowed_to_upload(&app_state.config, &auth_event.pubkey) {
+        return Err((StatusCode::UNAUTHORIZED, e.to_string()));
+    }
+
+    // Get the value of the Content-Length header, and check it against the allowed size.
+    let content_length = headers
+        .get("Content-Length")
+        .map(|v| v.to_str().unwrap_or_default().to_string())
+        .and_then(|v| v.parse::<u64>().ok());
+    let content_length = if let Some(val) = content_length {
+        val
+    } else {
+        return Err((StatusCode::BAD_REQUEST, "Missing or invalid Content-Length header".to_string()));
+    };
+    let blob_size_in_mb = bytes_to_mb(content_length as f64);
+    if blob_size_in_mb > app_state.config.upload.max_size {
+        return Err((StatusCode::BAD_REQUEST, format!(
+                "Blob size is {} MB, max upload size is {}",
```

Comment:
nitpick: Should we add "MB" behind the max upload size as well?

Reply:
Yes that makes sense, will add